apps/emqx/src/emqx_channel.erl

```diff
@@ -2114,9 +2121,14 @@ will_delay_interval(WillMsg) ->
         0
     ).
 
-publish_will_msg(Msg) ->
-    _ = emqx_broker:publish(Msg),
-    ok.
+publish_will_msg(ClientInfo, Msg = #message{topic = Topic}) ->
+    case emqx_access_control:authorize(ClientInfo, publish, Topic) of
+        allow ->
+            _ = emqx_broker:publish(Msg),
+            ok;
+        deny ->
```

Comment:
Add a WARNING level ?tp here and use it in test case .

Reply:
i.e. we want to warn users about this deny.

Reply:
:heavy_check_mark: